### ISSUE TITLE
Chessboard is hard to unit test

### ISSUE THREAD

**Issue by johng42** [dgschwind/Games#2]
I was looking at this and a simple refactor could make this class much more unit-testable.

summary:
replace the chessboard .print method with a method that creates the string(StringBuilder) that will be printed
example:  public String getBoard() {}

add a new print method that prints the string 
Example ChessBoard.print(String board)

then update the while loop in Chess to

      while (true) { // TODO
            String board = chessBoard.getBoard();

            chessBoard.print(board);

then you can add unit tests to validate the String before it gets printed.

I have this stubbed out if you are interested.  I think I would need permissions to push a branch up as a pull request though.  You *might* be able to see this with the branch name == refactorBoardForTesting

**Comment by dgschwind** [comment]
Hi John,

Thank you for taking a look! How did you happen to find my Games repository?

I agree with your assertion, that part of printing or verifying board state is poorly tested right now and doesn't lend itself yet to being properly unit tested. I still haven't implemented the part to be able to determine if a given King is in check or checkmate. I have been away from the implementation for a couple of weeks and will be busy for a couple more. Let me think on this great point you have raised. There are likely other areas that do not yet lend themselves very well to acceptable levels of unit testing.

Thank you, Doug

**Comment by johng42** [comment]
Hi Doug,

How I found this repo:
I was looking around for some Java text based games (of all things) to try and get running on a raspberry pi – just to learn how to get a Java environment set up on it.  I stumbled across your repo as one of the top hits on github and started digging in.

I had a tiny amount of free time but couldn’t figure out how to actually share my code as a change.  Git is still a bit of a mystery to me…

Enjoy!
John

**Comment by dgschwind** [comment]
Hey John,

Ah, very good. Very cool that you found my repo, but as you can see my Chess implementation is not yet complete. While my Poker and Boggle implementations are complete, they don't really allow people to play. Those were more of an exercise for me to have fun with Domain Modeling choices and see how they worked out. I have a couple of weeks in front of me with the holidays and such, I will think on your point a bit more.

Ah, I see. My repo, while public, was created in default state. I haven't considered any committers as yet, but if you have a desire to contribute to my Chess game implementation so that you can do your Raspberry pi work, am sure we can could that. I would need to change permissions on the repo, haven't tried that in the past.

Git is pretty simple actually. After you clone my repo, here are some steps:

1. `git checkout -b _yourBranchName_`
2. `git push -u origin _yourBranchName_`

Step one creates a branch off master in your local version of the repository only. The second step, somewhat optional at this point, creates a branch of the same name off master in the remote version of the repository. I suspect to be able to do step two, I will have to add you as a committer or the like. After step two is complete though, any pushes you do pushes the change from your local repo to the remote in github. Your changes of course still only reside on your branch. I tend to work this way to backup my work.

So after you do step one and two, make all the changes to the source files you feel are necessary, and you can do as many git commits as you like. Those commands typically look like:

1. `git commit -m 'Some text indicating what you changed and why'`
2. `git push`

After the `git push` command, your changes are present in your branch in your local copy of the repo as well as the remote repo.

Once you are happy with all of your changes and you wish to have those changes incorporated into `master`, you use the github UI to create a Pull Request (PR) that requests to promote the changes on _yourBranchName_ to `master`. Once I approve those changes, there should be a merge to master kind of button on the github UI, and when clicked, initiates getting the changes in the PR merged to `master`.

However, before we do all that, let me think on improving the unit testing of Chess. The unit tests for it are admittedly fairly sparse right at the moment.

Thanks, Doug

**Comment by dgschwind** [comment]
Hey John,

I have pushed some changes that address your concern. The print() method on ChessBoard now returns type String[] and no longer references System.out. I have added some assertions in ChessBoardTest to verify the state of the board is as expected in a few cases.

The overall implementation of Chess is not complete of course. I still need to add proper support for King being in check, and checkmate, as well as adding more testing of board state and of allowed movement of pieces mid game.

In any case, if there are other changes you would like to see in place to help you get an example set up on your Raspberry pi, do let me know.

Cheers, Doug
